### DISCUSSION TITLE
Using SML option "=so1" to correct power flow direction

### DISCUSSION THREAD

**Discussion by kricky** [arendst/Tasmota#17807] [Q&A]
Hello,
I have a problem decoding SML info from Tasmota 12.3.1 (compiled with SML_M and SCRIPT but withaout RULES).
My grid meter is a bidirectional EMH eHZ-H with energy import and export metering on OBIS 1.8.0 and 2.8.0 working well with Tasmota. The problem is with the current power value on OBIS 15.7.0 that is only send as absolute value. The direction of power flow is encoded in the status field of both the OBIS 1.8.0 and 2.8.0 telegrams (i.e. the ones sending the energy values for import and export in their value fields).
I struggle with the usage of the =so1 option to decode the status bit for inverting the sign of the power value.
My current script looks like this and works well but only gives the absolute power value without a sign for direction:
>D
>B
->sensor53 r
>M 1
+1,3,s,0,9600,
1,77070100010800ff@1000,Netzbezug,kWh,Grid_Total_Energy_Import,3
1,77070100020800ff@1000,Netzeinspeisung,kWh,Grid_Total_Energy_Export,3
1,770701000f0700ff@1,Netzleistung,W,Grid_Current_Power,17
#
 I tried to add "1,=so1,00010800,63,5,63,5,000f0700" but it did not result in a negative value when my PV produced more energy than used and exported it to the grid. 
The SML telegram looks like this when energy is purchased 
12:27:05.405 : 77 07 81 81 c7 82 03 ff 01 01 01 01 04 45 4d 48 01 
12:27:05.430 : 77 07 01 00 00 00 09 ff 01 01 01 01 0b 06 45 4d 48 01 02 71 5a 75 28 01 
12:27:05.453 : 77 07 01 00 01 08 00 ff 63 01 82 01 62 1e 52 ff 56 00 10 6e fd b9 01 
12:27:05.479 : 77 07 01 00 02 08 00 ff 63 01 82 01 62 1e 52 ff 56 00 18 f5 fe f2 01 
12:27:05.499 : 77 07 01 00 01 08 01 ff 01 01 62 1e 52 ff 56 00 10 6e fd b9 01 
12:27:05.521 : 77 07 01 00 02 08 01 ff 01 01 62 1e 52 ff 56 00 18 f5 fe f2 01 
12:27:05.543 : 77 07 01 00 01 08 02 ff 01 01 62 1e 52 ff 56 00 00 00 00 00 01 
12:27:05.565 : 77 07 01 00 0f 07 00 ff 01 01 62 1b 52 ff 55 00 00 03 47 01 
... and like this, when energy is fed into the grid:
12:27:23.024 : 77 07 81 81 c7 82 03 ff 01 01 01 01 04 45 4d 48 01 
12:27:23.049 : 77 07 01 00 00 00 09 ff 01 01 01 01 0b 06 45 4d 48 01 02 71 5a 75 28 01 
12:27:23.073 : 77 07 01 00 01 08 00 ff 63 01 a2 01 62 1e 52 ff 56 00 10 6e fd b9 01 
12:27:23.096 : 77 07 01 00 02 08 00 ff 63 01 a2 01 62 1e 52 ff 56 00 18 f5 fe f5 01 
12:27:23.119 : 77 07 01 00 01 08 01 ff 01 01 62 1e 52 ff 56 00 10 6e fd b9 01 
12:27:23.140 : 77 07 01 00 02 08 01 ff 01 01 62 1e 52 ff 56 00 18 f5 fe f5 01 
12:27:23.162 : 77 07 01 00 01 08 02 ff 01 01 62 1e 52 ff 56 00 00 00 00 00 01 
12:27:23.183 : 77 07 01 00 0f 07 00 ff 01 01 62 1b 52 ff 55 00 00 03 bd 01 
The 63 01 xx is the status information I need to decode. When changing from import to export the last byte changes from 82h--> 80h --> a0h --> a2h
Any idea how I could get my absolute power value corrected to a signed power value? 
Thanks a lot in advance :-)

**Comment by gemu2015** [comment]
63 01 82
63 01 a2
so bit 5 changes
"1,=so1,00010800,63,5,63,5,000f0700"
in theory your calculation of so1 is correct and it should work.

USE_SML_SPECOPT is only default enabled recently. if you don't use the latest dev, did you enable USE_SML_SPECOPT ?

**Comment by bejoup** [comment]
@kricky i have the Same smartmeter and Problem.  Could you solve it? An which tasmota Version are you using? Is this option USE_SML_SPECOPT by default activated in Tasmota 13.0.0

i have posted my way to a Solution https://github.com/arendst/Tasmota/issues/8001 
But i have a Problem with my hichi, because it won‘t understand the Special command ☹️

**Reply by bejoup** [reply]
Problem solved!

**Reply by kricky** [reply]
@bejoup it did not work for me but I cannot remember exactly the problem i encountered because it was quite a while ago. I am planning to use the winter time to work on the project again. Can you share how you solved it?

**Reply by bejoup** [reply]
I postet the Solution here: 
https://github.com/arendst/Tasmota/issues/8001
Have fun!
